### ISSUE TITLE
Quantizing Gradients - Meaning of max0() operator in DoReFa v2 paper?

### ISSUE THREAD

**Issue by the-bobo** [ppwwyyxx/tensorpack#30]
Thank you for your help so far. 

(1) In section 2.5 on quantizing gradients you use an operator called _max<sub>0</sub>_ but do not define it. I did not find a definition in the XNOR or BNN papers either. What does this operator do? How is it different from the regular `max()` operator?

(2) Second, you say that _dr_ / 2max<sub>0</sub>(|_dr_|) + 1/2 is an affine transform to map the gradient into [0,1], but it seems like in your code you apply an additional step to manually clip the values. Why do you need this additional step?

Code: https://github.com/ppwwyyxx/tensorpack/blob/master/examples/DoReFa-Net/dorefa.py

``` python
 def grad_fg(op, x):
            rank = x.get_shape().ndims
            assert rank is not None
            maxx = tf.reduce_max(tf.abs(x), list(range(1,rank)), keep_dims=True)
            x = x / maxx
            n = float(2**bitG-1)
            x = x * 0.5 + 0.5 + tf.random_uniform(
                    tf.shape(x), minval=-0.5/n, maxval=0.5/n)
            x = tf.clip_by_value(x, 0.0, 1.0) # this is the extra step not in the paper
            x = quantize(x, bitG) - 0.5
            return x * maxx * 2
```

(3) I am also having trouble understanding this line, could you please explain? - `maxx = tf.reduce_max(tf.abs(x), list(range(1,rank)), keep_dims=True)`. 

It seems like `list(range(1,rank))` is somehow related to your statement that "Here _dr_ = _∂c/∂r_ is the back-propagated gradient of the output _r_ of some layer, and the maximum is taken over all axis of the gradient tensor _dr_ except for the mini-batch axis (therefore each instance in a mini-batch will have its own scaling factor)", but I do not understand this sentence either. Thank you for your help!

**Comment by ppwwyyxx** [comment]
Hi,
1) 3) __the maximum is taken over all axis of the gradient tensor *dr* except for the mini-batch axis.__ This is  the definition of `max_0`, and it is the line:
` maxx = tf.reduce_max(tf.abs(x), list(range(1,rank)), keep_dims=True)`
which takes the maximum on all axes except the first (batch) axis.

2) In the formulation of the paper we need the stochastic noise parameter in the open interval (-0.5/(2^k-1), 0.5/(2^k-1)), but in tensorflow, `tf.random_uniform` return a random number in the left-close-right-open interval.
When the left endpoint of the interval is reached, we'll need to quantize `-0.5/(2^k-1)`, and this may lead to a negative output, depending on whether round(-0.5) is implemented to be 0 or -1. Although this scenario is unlikely to happen, I clipped the value here only for safety.

**Comment by the-bobo** [comment]
Thanks - I've been experimenting with `maxx = tf.reduce_max(tf.abs(x), list(range(1,rank)), keep_dims=True)` to understand what it does. What do you mean by the "batch" axis?

**Comment by ppwwyyxx** [comment]
Don't know which line are you talking about, but in general the code should work for any n-dimensional tensor, where the first dimension is the mini-batch dimension.

**Comment by the-bobo** [comment]
Thanks - I do not know what you mean when you say "the first dimension is the mini-batch dimension." I'm familiar with mini-batch but not with TensorFlow.

**Comment by ppwwyyxx** [comment]
Usually when a mini-batch of B images of shape [H,W,C] is fed into the network for training, they are stacked to form one big tensor of shape [B,H,W,C] and get processed together. This is the case for most frameworks.
It's the same case for all the intermediate layer inputs/outputs. The first dimension is usually the batch dimension.

**Comment by the-bobo** [comment]
Oh I see, if I understand correctly B is just the size of the mini-batch, so "the mini-batch dimension" just means the number of training samples in the mini-batch. Thank you.

I'm not very familiar with convolutional nets. I'm currently trying to reproduce your team's results on a simple feedforward network. Is the C in [B,H,W,C] a vector with values for the channels red, green, blue?  (With H and W being height and width?)

**Comment by ppwwyyxx** [comment]
Yes C is for channels.

**Comment by the-bobo** [comment]
Thanks for your help. I'd like to confirm that I understand your gradient quantization method.

Let's say I was implementing DoReFa algorithm on a simple feedforward network. The output layer just outputs a single vector: a list of scalar values: O = [o<sub>1</sub>, o<sub>2</sub>, o<sub>3</sub>, o<sub>4</sub>, ..., o<sub>n</sub>].

I compute the gradients at the output layer: G = [g<sub>1</sub>, g<sub>2</sub>, g<sub>3</sub>, g<sub>4</sub>, ..., g<sub>n</sub>]

To produce the quantized gradients g<sup>b</sup><sub>j</sub> I then do the following (as in Equation (12)):
∀g<sub>j</sub> ∈ G, g<sup>b</sup><sub>j</sub> = 2*max(|G|) * ( quantize<sub>k</sub> [g<sub>j</sub> / 2 * max(|G|) + ½ + N(_k_)] – ½ ).

Where `max(|G|)` is the element in G with the maximum absolute value.  So if G := [0, 4, -2, 1, -5], `max(|G|)` returns `5`.

**Comment by ppwwyyxx** [comment]
Yes that's right. Because it looks like there is no mini-batch (or in other words, a batch-size of 1).

**Comment by the-bobo** [comment]
Thanks, if I did use a mini-batch with size > 1, say m = 4, then it seems like all that is necessary is to sum and average the quantized gradients. So in other words, if m= 4 then we have 4 training inputs, x<sub>1</sub>, x<sub>2</sub>, x<sub>3</sub>, x<sub>4</sub> per mini-batch, and the gradient is: 1/4 * Σ<sub>t=1 to m</sub> [ G<sub>t</sub><sup>b</sup> ], where G<sub>t</sub> is the output gradient vector from training input x<sub>t</sub>, and G<sub>t</sub><sup>b</sup> is the quantized gradient vector from training input x<sub>t</sub>.

**Comment by ppwwyyxx** [comment]
Apart from the quantization it's just standard back-propagation. For example if your cost function is the average cost of all samples then the gradient w.r.t some weight would also be averaged across all samples.

**Comment by the-bobo** [comment]
Sorry - I should have specified, where mini-batch size m > 1 I was implying that stochastic gradient descent would be used, where an estimate of the true gradient ∇C = 1/n * Σ<sub>x</sub> C<sub>x</sub> ∀x ∈ X is given instead by ∇C ≈ 1/m * Σ<sub>t=1 to m</sub>( ∇C<sub>X<sub>t</sub></sub> )

where _X_ is the set of all training samples and _C_ is the cost function, and _m_ is the size of a single mini-batch.

> For example if your cost function is the average cost of all samples then the gradient w.r.t some weight would also be averaged across all samples.

I was under the impression that [all cost functions](http://stats.stackexchange.com/questions/154879/a-list-of-cost-functions-used-in-neural-networks-alongside-applications) must be able to be written as an average over cost functions _C<sub>x</sub>_ for individual training samples _x_ ∈ X. Is this incorrect?

**Comment by ppwwyyxx** [comment]
Yeah I had that impression as well but I'm not sure it has to be the case.. 
Anyway, if cost function is an average then averaging the gradient is necessary by chain-rule.

**Comment by the-bobo** [comment]
I don't remember you mentioning biases in the paper - do you have any method for quantizing biases? Or do you simply imagine a single bias unit connected to all neurons in all layers, and train the weight for those connections?

**Comment by ppwwyyxx** [comment]
Bias are not of a specific interest because they don't involve expensive computation, and presumably similar quantization methods would work for them as well. In our work we use batch normalization so the model doesn't contain bias term for those quantized layers.

**Comment by the-bobo** [comment]
Confirming one more detail. On step 13 of Algorithm 1 in DoReFa v2 paper, you say:

> 13: g<sub>a<sub>k-1</sub></sub>←`backward_input`(g<sup>b</sup><sub>a<sub>k</sub></sub>, W<sup>b</sup><sub>k</sub>)

Does this mean that we have to re-quantize the gradients g<sub>a<sub>k-1</sub></sub> using _f_<sup>G</sup><sub>γ</sub>() before backpropagating back another layer?

**Comment by the-bobo** [comment]
Another question - when we implement `backward_input()`, we end up doing the following:
dC/da<sub>L</sub> ⊙ σ’(z⃗<sub>L</sub>) = δ⃗<sub>L</sub>. 
(where dC/da<sub>L</sub> = g<sub>a<sub>L</sub></sub> from step 9 of DoReFa v2 Algorithm 1, and σ’(z⃗<sub>L</sub>) is the derivative of the activation function σ() evaluated over the weighted inputs z⃗<sub>L</sub>)

This gives us the error vector δ⃗<sub>L</sub> from which we can use `backward_weight()` to get the gradients with respect to the weight matrix.

The questions: 
1) What bit-width do you use for the components of the vector σ’(z⃗<sub>L</sub>)?
2) What bit-width do you use for the vector z⃗<sub>L</sub> (this may already be answered [here](https://github.com/ppwwyyxx/tensorpack/issues/27#issuecomment-250816137) as a bitwidth of (M+K) where M and K are the bit-widths of the multiplicands used to generate the components of z⃗<sub>L</sub>)

**Comment by ppwwyyxx** [comment]
Quantization is done in step 12 already. Not step 13.

`backward_input` doesn't involve anything with the activation function because it is handled in step 11.
You can use whatever bitwidth large enough to store these intermediate values because they only get involved in cheap operations.

**Comment by the-bobo** [comment]
Thanks my mistake - steps 12 and 13 are part of a loop...

**Comment by the-bobo** [comment]
I think I misunderstood `backward_input` and `backward_weight`. Can you please explain what steps 13 and 14 do? It is unclear how they operate within the loop started in step 10 (Algorithm 1, DoReFa v2 paper).

When backpropagating the following things need to happen:
1) Compute the output layer's gradient: this is dC/da<sub>L</sub>, the derivative of the cost function with respect to the final layer (layer L)'s output activations. 

This looks like step 9 in Algorithm 1: "Compute g<sub>a<sub>L</sub></sub> = ∂C/∂a<sub>L</sub> knowing a<sub>L</sub> and a*."

2) Feed this gradient through the derivative of the activation function, to get the error δ<sub>L</sub> of the output layer: δ<sub>L</sub> = g<sub>a<sub>L</sub></sub>⊙σ’(z<sub>L</sub>), where σ’() is the derivative of the activation function, z<sub>L</sub> are the weighted inputs to the last layer L, and ⊙ is the Hadamard product (elementwise product).

This looks like step 11 in Algorithm 1: "Back-propagate _g<sub>a<sub>k</sub></sub>_ through activation function _h_."

3) Backpropagate the error δ<sub>L</sub> ∀ ℓ < L. δ<sub>ℓ</sub> = ((w<sub>ℓ+1</sub>)<sup>T</sup> · δ<sub>ℓ+1</sub>) ⊙σ’(z<sub>ℓ</sub>).

This might be step 13? Except the output of step 13 is _non-quantized_ (note there is no <sup>b</sup>, the output is simply _g<sub>a<sub>k-1</sub></sub>_).

4) Apply the update to the weights: w<sub>ℓ</sub> → w<sub>ℓ</sub> - (η/m) * (a<sub>ℓ-1</sub> · δ<sub>ℓ</sub>), where a<sub>ℓ-1</sub> is the activation vector from layer ℓ-1 and w<sub>ℓ</sub> is the weight matrix at layer ℓ.

Questions:
Q1) In the four steps outlined above for backprop, when do we apply the quantization routine in Step 12 from Algorithm 1?
Q2) What is the meaning of `backward_input()` - is it one of the four steps above or something else?
Q3) What is the meaning of `backward_weight()` - is it one of the four steps above or something else?
Q4) Why is the output of step 13's `backward_input()` non-quantized?

**Comment by ppwwyyxx** [comment]
`backwad_input` computes the gradient of C w.r.t.  a_{L-1} (or L-2, L-3,...), i.e. the output of previous layer. So the next loop iteration can use it.
`backward_weight` computes the gradient w.r.t. W_l (or L-1, L-2,...).

Yes in step 11 we just assume this is a trivial element-wise operation so we didn't introduce new notations to distinguish what's before the activation and what's after. (or "gradients" and "errors" in your words). So g_{a_k} is a variable that holds the gradient w.r.t. the output of layer k either before or after activation.

**Comment by JINwonLEE** [comment]
I have some questions in getting gradient quantization. In the paper, it said that 
"The above function first applies an affine transform on the gradient, to map it into [0, 1], and then inverts the transform after quantization."

In this sentence, what is "inverts" mean?

I can't get the points.. Can I get some advice?

**Comment by ppwwyyxx** [comment]
It means applying the [inverse function](https://www.wikiwand.com/en/Inverse_function) of the transform.

**Comment by JINwonLEE** [comment]
Oh, I get it.. However why do we have to invert the transform? Is there any special reason? Also I can't get the intention of multiplying max_0(|dr|) in the front. Can you explain little bit in detail?

**Comment by ppwwyyxx** [comment]
>  Is there any special reason? 

To keep things in the same range.

> Also I can't get the intention of multiplying max_0(|dr|) in the front. Can you explain little bit in detail?

That's part of the inverse function.

**Comment by JINwonLEE** [comment]
Thanks a lot...!!! I got the all equation.

Sorry but I had little questions about N(k), 

Noise function N(k).. In the paper, 
 1. Is there any guarantee that dr/(2 * max_0(|dr|)) + 1/2 + N(k) < 1? (Becuase quantization should be in [0,1])

2. Can you give me some explanation about specific affect of noise function(N(k)) in gradient quantization? (For example, the performance becomes better because of some aspects... of noise function..)

**Comment by JINwonLEE** [comment]
Oh I didn't see the code yet. Thanks a lot. 

For 2., I should see the paper more in detail.

**Comment by JINwonLEE** [comment]
While reading the paper, I have something curious. 

In the contribution, paper said 
"DoReFa-Net can use bit convolution kernels to accelerate both the forward pass and the backward pass of the training process."

In this sentence, what is bit convolution kernel? Is it implemented by authors or just using exist thing?
I thought that it is implemented by authors, but I want to make sure and if it was implemented then is it open source?

**Comment by ppwwyyxx** [comment]
They are implemented and not open source.
btw recently caffe2 open source some relevant low-precision kernels: https://github.com/caffe2/caffe2/tree/master/caffe2/mobile/contrib/ulp2.

**Comment by JINwonLEE** [comment]
Oh Thanks for reference. 

I had some minor questions in the experiment.

In the ImageNet experiment, there are case which is initialized. In the paper, it said that "training has been initiailized with a 32-bit model."

As I understand, the case "initialized" means that even though they are using 1-bit weight, activation and 32-bit gradients, they initialized all the 3 values(weight, activation, gradients) in 32-bit, different from other case which is initialized in lower than 32 bit. Am I in right track? 

If not, can I get some advice?

**Comment by ppwwyyxx** [comment]
You can only initialize the weights when training a neural network.

**Comment by JINwonLEE** [comment]
Ah.. you mean that the case "initialized" is that they initialized weights with a 32bit and the other case is just randomly chosen by W bits. Is it right?

**Comment by ppwwyyxx** [comment]
Initialized means initializing with weights from pretrained 32bit model. Otherwise they're randomly initialized.

**Comment by JINwonLEE** [comment]
Ah.. now I understand what the table means, The cases "initialized" are start from initialized weight with pretrained 32bit model and the others are initialized with randomized values!!!
